04-onion-routing.md

```diff
@@ -991,46 +1000,102 @@ Note that these error messages are not reliable, as they are not placed on-chain
+node supports this format by observing the node feature
+`option_attributable_error`.
+
+By choosing a structure for the attributable errors, the origin node is making a
+trade-off between failure message size and privacy. For maximum privacy,
+`hop_count` should be set to the global maximum of 27. A lower value reveals a
+lower upper bound for the route length.
+
+## Erring node
+
+The node generating the error message builds a return packet
 consisting of the following fields:
 
 1. data:
-   * [`32*byte`:`hmac`]
    * [`u16`:`failure_len`]
    * [`failure_len*byte`:`failuremsg`]
    * [`u16`:`pad_len`]
    * [`pad_len*byte`:`pad`]
+   * [`hop_count*(fixed_payload_len+1)*byte`:`payloads`]
+   * [`total_hmacs*32*byte`:`hmacs`]
 
-Where `hmac` is an HMAC authenticating the remainder of the packet, with a key
-generated using the above process, with key type `um`, `failuremsg` as defined
-below, and `pad` as the extra bytes used to conceal length.
-
-The erring node then generates a new key, using the key type `ammag`.
-This key is then used to generate a pseudo-random stream, which is in turn
-applied to the packet using `XOR`.
+Where `failuremsg` is defined below, and `pad` are the extra bytes used to
+conceal length.
 
-The obfuscation step is repeated by every hop along the return path.
-Upon receiving a return packet, each hop generates its `ammag`, generates the
-pseudo-random byte stream, and applies the result to the return packet before
-return-forwarding it.
+The field `payloads` contains a per-hop payload. The erring node puts its
+payload at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum number of hops in a route (`hop_count`) and the
+payload size (`fixed_payload_len`). An extra payload byte is added to the size
+signaled in the `attributable_error` onion tlv to indicate the source of the
+payload.
 
-The origin node is able to detect that it's the intended final recipient of the
-return message, because of course, it was the originator of the corresponding
-forward packet.
-When an origin node receives an error message matching a transfer it initiated
-(i.e. it cannot return-forward the error any further) it generates the `ammag`
-and `um` keys for each hop in the route.
-It then iteratively decrypts the error message, using each hop's `ammag`
-key, and computes the HMAC, using each hop's `um` key.
-The origin node can detect the sender of the error message by matching the
-`hmac` field with the computed HMAC.
+Currently `fixed_payload_len` is fixed at 7 (excludes `payload_source`) to just
+fit the hold time in tlv format. The field exists for future extensibility.
 
-The association between the forward and return packets is handled outside of
-this onion routing protocol, e.g. via association with an HTLC in a payment
-channel.
+The per-hop payload consists of the following fields:
+1. data:
+   * [`byte`:`payload_source`]
+   * [`bigsize`:`tlv_stream_len`]
```

Comment:
As mentioned in https://github.com/lightning/bolts/pull/1044/files#r1234361494, we might just fix the payload length to 4 bytes excluding `payload_source`. No tlv, just a 32-bit hold time in milliseconds.